File: src/main/java/com/meada/profiles/barbearia/BarberAppointmentStatus.java
package com.meada.profiles.barbearia;

import java.util.Arrays;
import java.util.Optional;
import java.util.Set;

/**
 * Status de um agendamento de barbearia (camada 8.1) — clone do {@code SalonAppointmentStatus}:
 * <pre>
 *   agendado   → confirmado, cancelado
 *   confirmado → realizado, cancelado, falta
 *   realizado  → (terminal)
 *   cancelado  → (terminal)
 *   falta      → (terminal)
 * </pre>
 * Transição inválida → 409 invalid_status_transition no controller.
 *
 * <p>Espelhado 1:1 por {@code frontend/profiles/barbearia/barber-appointment-status.ts}
 * (BarberAppointmentStatusParityTest garante a paridade Java↔TS).
 */
public enum BarberAppointmentStatus {
    AGENDADO("agendado", "Agendado"),
    CONFIRMADO("confirmado", "Confirmado"),
    REALIZADO("realizado", "Realizado"),
    CANCELADO("cancelado", "Cancelado"),
    FALTA("falta", "Falta");

    private final String id;
    private final String label;

    BarberAppointmentStatus(String id, String label) {
        this.id = id;
        this.label = label;
    }

    public String id() {
        return id;
    }

    public String label() {
        return label;
    }

    public static Optional<BarberAppointmentStatus> fromId(String id) {
        if (id == null) {
            return Optional.empty();
        }
        return Arrays.stream(values()).filter(s -> s.id.equals(id)).findFirst();
    }

    /** Transições permitidas a partir deste status. */
    public Set<BarberAppointmentStatus> allowedNext() {
        return switch (this) {
            case AGENDADO -> Set.of(CONFIRMADO, CANCELADO);
            case CONFIRMADO -> Set.of(REALIZADO, CANCELADO, FALTA);
            case REALIZADO, CANCELADO, FALTA -> Set.of();
        };
    }

    public boolean canTransitionTo(BarberAppointmentStatus next) {
        return allowedNext().contains(next);
    }

    /**
     * Texto fixo da notificação outbound ao ENTRAR neste status. null = não notifica. {@code cancelado}
     * avisa; {@code confirmado} é parametrizado — ver {@link #notificationText(String, String, String)}.
     */
    public String notificationText() {
        return switch (this) {
            case CANCELADO -> "Seu horário foi cancelado. Pra remarcar, é só me chamar.";
            case CONFIRMADO, AGENDADO, REALIZADO, FALTA -> null;
        };
    }

    /**
     * Texto da notificação de CONFIRMAÇÃO, que depende dos dados do agendamento. Para os demais
     * status, devolve {@link #notificationText()}. Texto defensivo (sem promessa de resultado de corte).
     */
    public String notificationText(String dateLabel, String timeLabel, String barberName) {
        if (this == CONFIRMADO) {
            return "Seu horário foi confirmado pra " + dateLabel + " às " + timeLabel
                + " com " + barberName + ". Te esperamos!";
        }
        return notificationText();
    }
}

File: frontend/profiles/barbearia/barber-appointment-status.ts
/**
 * Status de um agendamento do perfil barbearia (camada 8.1) — espelho 1:1 de
 * src/main/java/com/meada/profiles/barbearia/BarberAppointmentStatus.java.
 *
 * O BarberAppointmentStatusParityTest (backend) garante que os ids aqui e no enum Java nunca
 * divergem. A CHECK constraint de barber_appointments.status (migration 43) trava os mesmos ids.
 *
 * Transições:
 *   agendado   → confirmado, cancelado
 *   confirmado → realizado, cancelado, falta
 *   realizado/cancelado/falta → (terminal)
 */
export const BARBER_APPOINTMENT_STATUSES = [
  { id: 'agendado', label: 'Agendado' },
  { id: 'confirmado', label: 'Confirmado' },
  { id: 'realizado', label: 'Realizado' },
  { id: 'cancelado', label: 'Cancelado' },
  { id: 'falta', label: 'Falta' },
] as const

export type BarberAppointmentStatus = (typeof BARBER_APPOINTMENT_STATUSES)[number]
export type BarberAppointmentStatusId = BarberAppointmentStatus['id']

/** Transições permitidas a partir de cada status (espelha BarberAppointmentStatus.allowedNext). */
export const ALLOWED_NEXT: Record<BarberAppointmentStatusId, BarberAppointmentStatusId[]> = {
  agendado: ['confirmado', 'cancelado'],
  confirmado: ['realizado', 'cancelado', 'falta'],
  realizado: [],
  cancelado: [],
  falta: [],
}

/** Rótulo pt-BR de um status (fallback: o próprio id se desconhecido). */
export function statusLabel(id: string): string {
  return BARBER_APPOINTMENT_STATUSES.find((s) => s.id === id)?.label ?? id
}
